refactor lazy init to device-agnostic

guangyey opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #118613
* #118528
* #118523
* #118091
* #117734
* #117619
* #117611
* #116869
* __->__ #118846

# Motivation
This PR intends to extend `cuda_lazy_init` to `device_lazy_init` which is a device-agnostic API that can support any backend. And change `maybe_initialize_cuda` to `maybe_initialize_device` to support lazy initialization for CUDA while maintaining scalability.

# Design
We maintain a flag for each backend to manage the lazy initialization state separately.

# Additional Context
No need more UTs.

guangyey commented:
Hi @malfet, I seperate  `device_lazy_init` to this PR. Please help review.

guangyey committed:
refactor lazy init to device-agnostic

[ghstack-poisoned]

guangyey committed:
Update on "refactor lazy init to device-agnostic"




[ghstack-poisoned]

guangyey committed:
Update on "refactor lazy init to device-agnostic"


# Motivation
This PR intends to extend `cuda_lazy_init` to `device_lazy_init` which is a device-agnostic API that can support any backend. And change `maybe_initialize_cuda` to `maybe_initialize_device` to support lazy initialization for CUDA while maintaining scalability.

# Design
We maintain a flag for each backend to manage the lazy initialization state seperately.


[ghstack-poisoned]

guangyey committed:
Update on "refactor lazy init to device-agnostic"


# Motivation
This PR intends to extend `cuda_lazy_init` to `device_lazy_init` which is a device-agnostic API that can support any backend. And change `maybe_initialize_cuda` to `maybe_initialize_device` to support lazy initialization for CUDA while maintaining scalability.

# Design
We maintain a flag for each backend to manage the lazy initialization state seperately.


[ghstack-poisoned]

malfet commented on torch/csrc/utils/device_lazy_init.cpp:
```diff
@@ -0,0 +1,43 @@
+namespace torch::utils {
+namespace {
+
+// NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables,modernize-avoid-c-arrays,cppcoreguidelines-avoid-c-arrays)
+bool is_initialized[at::COMPILE_TIME_MAX_DEVICE_TYPES] = {false};
```
Please use std::array here, as the construct you wrote here will initialize only first element of the array, see https://godbolt.org
```suggestion
std::array<bool, at::COMPILE_TIME_MAX_DEVICE_TYPES>  is_initialized{};
```

guangyey replied:
Thanks for your suggestion. Done.

malfet reviewed (approved):
LGTM, but please fix `is_initialized` initialization

guangyey commented:
@torchbot merge

guangyey commented:
@pytorchbot merge

guangyey commented:
@pytorchbot merge

guangyey committed:
Update on "refactor lazy init to device-agnostic"


# Motivation
This PR intends to extend `cuda_lazy_init` to `device_lazy_init` which is a device-agnostic API that can support any backend. And change `maybe_initialize_cuda` to `maybe_initialize_device` to support lazy initialization for CUDA while maintaining scalability.

# Design
We maintain a flag for each backend to manage the lazy initialization state separately.

# Additional Context
No need more UTs.

[ghstack-poisoned]

guangyey committed:
Update on "refactor lazy init to device-agnostic"


# Motivation
This PR intends to extend `cuda_lazy_init` to `device_lazy_init` which is a device-agnostic API that can support any backend. And change `maybe_initialize_cuda` to `maybe_initialize_device` to support lazy initialization for CUDA while maintaining scalability.

# Design
We maintain a flag for each backend to manage the lazy initialization state separately.

# Additional Context
No need more UTs.

[ghstack-poisoned]

atalman commented:
@pytorchbot revert -c ghfirst -m "Failing, tests https://github.com/pytorch/torchdistx/blob/main/src/python/torchdistx/_C/fake.cc#L11 "
